src/libmeasurement_kit/engine.cpp

```diff
@@ -0,0 +1,480 @@
+// Part of Measurement Kit <https://measurement-kit.github.io/>.
+// Measurement Kit is free software under the BSD license. See AUTHORS
+// and LICENSE for more information on the copying conditions.
+
+#define MK_ENGINE_INTERNALS
+#include <measurement_kit/engine.h>
+
+#include <assert.h>
+#include <stdint.h>
+
+#include <algorithm>
+#include <atomic>
+#include <condition_variable>
+#include <deque>
+#include <exception>
+#include <future>
+#include <memory>
+#include <mutex>
+#include <set>
+#include <string>
+#include <sstream>
+#include <thread>
+#include <tuple>
+#include <type_traits>
+
+#include <measurement_kit/common/error.hpp>
+#include <measurement_kit/common/logger.hpp>
+#include <measurement_kit/common/nlohmann/json.hpp>
+#include <measurement_kit/common/reactor.hpp>
+#include <measurement_kit/common/shared_ptr.hpp>
+
+#include "src/libmeasurement_kit/nettests/runnable.hpp"
+
+namespace mk {
+namespace engine {
+
+// # Multi-thread stuff
+//
+// Comes first because it needs more careful handling.
+
+class Semaphore {
```

Comment:
Would this implementation be better? (My understanding is that a semaphore with only one active thread at a time is actually equivalent to a mutex):

```
diff --git a/src/libmeasurement_kit/engine.cpp b/src/libmeasurement_kit/engine.cpp
index 07f9604d..aebe1820 100644
--- a/src/libmeasurement_kit/engine.cpp
+++ b/src/libmeasurement_kit/engine.cpp
@@ -42,25 +42,18 @@ class Semaphore {
   public:
     Semaphore() = default;
 
-    void acquire() {
-        std::unique_lock<std::mutex> lock{mutex_};
-        cond_.wait(lock, [this]() { return !active_; });
-        active_ = true;
-    }
+    void acquire() { mutex_.lock(); }
 
-    void release() {
-        {
-            std::unique_lock<std::mutex> _{mutex_};
-            active_ = false;
-        }
-        cond_.notify_one();
-    }
+    void release() { mutex_.unlock(); }
+
+    Semaphore(const Semaphore &) = delete;
+    Semaphore &operator=(const Semaphore &) = delete;
+    Semaphore(Semaphore &&) = delete;
+    Semaphore &operator=(Semaphore &&) = delete;
 
     ~Semaphore() = default;
 
   private:
-    bool active_ = false;
-    std::condition_variable cond_;
     std::mutex mutex_;
 };
 
```

Reply:
(But, I wonder whether this will preserve ordering, I should check. Otherwise, another possible implementation is probably one where there is a singleton thread with a work queue. In the latter case, we will still use a condition variable, a mutex, a queue. Also, in such case, I think - but need to check - that we may need to register an atexit handler to quit the thread.)

Reply:
> But, I wonder whether this will preserve ordering, I should check.

`Notify_one()` does not guarantee that order is preserved. The documentation says ["unblocks one of the waiting threads"](http://en.cppreference.com/w/cpp/thread/condition_variable/notify_one). So, it does not matter whether using a mutex preserves ordering or not, because we were not preserving it already.

Do we want to enforce strict ordering at this level? Probably, we don't.

Rationale:

1. I would say that, if an app wants to run tests in a _specific_ order, it should run them in a specific order. Since the API is _blocking_, this is the most natural outcome you get if you use a single worker thread for running tests. Running tests in parallel, entails more work. It would require you to spawn several threads and to arrange to concurrently update the UI for many tests. The current OONI probe for {iOS,Android} does that. But the new version will not, since we do not want to encourage users to _attempt_ to run tests in parallel.

2. The way I see it, this `Semaphore` is just an internal safeguard to prevent the app developer from mis-using the library. As such, I would say it's okay to make its implementation _as simple as possible_ (hence a single mutex). We probably want to specify in the documentation that, in the event in which one attempts to run tests in parallel, they will be serialised, and the order will not be preserved. Do we also want to issue a warning in such case?

- - -

BTW, I also checked whether perhaps, _at the implementation level_, we can rely on the system dependent code to actually enforce FIFO behaviour. To this end, I looked into the sources of [llvm-mirror/libcxx](https://github.com/llvm-mirror/libcxx). Long story short: on Windows a SRW lock is used, which [does not preserve order](https://msdn.microsoft.com/en-us/library/windows/desktop/aa904937(v=vs.85).aspx). On Unix, pthread_mutex_unlock is used, where [the order in which the lock is released depends on the scheduling policy](http://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_mutex_lock.html), which is not totally clear to me (I think it means that it depends on the kernel scheduler, tho, since there is no reference to any other man page).

Reply:
So, I also pinged @lorenzoPrimi and @hellais on this. It's not clear whether we want to enforce FIFO order or not. But, in any case, it seems to me that, if we don't want FIFO order, then the implementation using a single mutex is functionally equivalent to the existing one and simpler.